### PR TITLE
[Inductor] Enable VecMask store

### PR DESCRIPTION
**Pull request by leslie-fang-intel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #123710
* #123512

**Summary**
Enable the vectorization of store with `bool` dtype. 

**Test Plan**
```
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_decomposed_fake_quant_per_channel 
```

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/codegen/cpp.py [hunk 1]
```diff
@@ -2881,6 +2881,7 @@ def __init__(self, args, num_threads, tiling_factor, tiling_idx=-1):
             torch.int8,
             torch.int32,
             torch.int64,
+            torch.bool,
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp.py:2884]: with this change, we can use one list `supported_dtypes` for all such checks since the lists are all the same now. Can we define one list and used by all the checks in the code?

File: test/inductor/test_cpu_repro.py [hunk 2]
```diff
@@ -1825,6 +1825,8 @@ def run_compile_decomposed_fake_quant(
         self.assertEqual(input_grad_aten_eager, input_grad)
         self.assertEqual(input_grad_decomp_eager, input_grad)
         self.assertEqual(input_grad[1, 2, 3, 4], torch.tensor(0.0))
+        # For forward and backward kernel
+        check_metrics_vec_kernel_count(2)
```
Comment (jgong5) [inline test/inductor/test_cpu_repro.py:1829]: Is this a related change? Do we have UT on vectorized bool?

File: aten/src/ATen/cpu/vec/vec_mask.h [hunk 3]
```diff
@@ -134,6 +134,14 @@ class VecMask {
     return VectorizedN<T, N>(VectorizedN<T, N>::loadu(mask));
   }
 
+  void store(bool* b, int count) {
+    TORCH_CHECK(count == Vectorized<T>::size() * N, "Expect same number of elements in VecMask::store.");
+    TORCH_CHECK(count <= Vectorized<bool>::size(), "Expect the number is less or equal to Vectorized<bool>.");
+    Vectorized<bool> res = this->to<bool, 1>();
```
Comment (jgong5) [inline aten/src/ATen/cpu/vec/vec_mask.h:140]:
```c++
constexpr int L = (VectorizedN<T, N>::size() + Vectorized<bool>::size() - 1)/ Vectorized<bool>::size();
auto res = this->to<bool, L>();
res.store(b, count);
```

File: aten/src/ATen/cpu/vec/vec_mask.h [hunk 4]
```diff
@@ -134,6 +134,14 @@ class VecMask {
     return VectorizedN<T, N>(VectorizedN<T, N>::loadu(mask));
   }
 
+  void store(bool* b, int count) {
+    TORCH_CHECK(count == Vectorized<T>::size() * N, "Expect same number of elements in VecMask::store.");
+    TORCH_CHECK(count <= Vectorized<bool>::size(), "Expect the number is less or equal to Vectorized<bool>.");
```
Comment (jgong5) [inline aten/src/ATen/cpu/vec/vec_mask.h:139]: No need to explicit assert here? It should be caller's responsibility to guarantee the count is valid?

File: aten/src/ATen/cpu/vec/vec_mask.h [hunk 5]
```diff
@@ -134,6 +134,14 @@ class VecMask {
     return VectorizedN<T, N>(VectorizedN<T, N>::loadu(mask));
   }
 
+  void store(bool* b, int count) {
```
Comment (jgong5) [inline aten/src/ATen/cpu/vec/vec_mask.h:137]:
```suggestion
  void store_mask(bool* b, int count) {
```

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: torch/_inductor/codegen/cpp.py (hunk 1 above)
Comment (leslie-fang-intel) [reply torch/_inductor/codegen/cpp.py:2884]: Thanks for the comment. Merge into one list of `supported_dtypes`.

File: test/inductor/test_cpu_repro.py (hunk 2 above)
Comment (leslie-fang-intel) [reply test/inductor/test_cpu_repro.py:1829]: Yean. We have a UT `python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_decomposed_fake_quant_per_channel `. Previously, the forward graph has been disabled of vectorization due to the store of bool. And after this PR, we can vectorize the forward graph now.

File: aten/src/ATen/cpu/vec/vec_mask.h (hunk 3 above)
Comment (leslie-fang-intel) [reply aten/src/ATen/cpu/vec/vec_mask.h:140]: Thanks for the comments.

File: aten/src/ATen/cpu/vec/vec_mask.h (hunk 4 above)
Comment (leslie-fang-intel) [reply aten/src/ATen/cpu/vec/vec_mask.h]: Sure and removed.

File: torch/_inductor/codegen/cpp.py [hunk 6]
```diff
@@ -2902,7 +2892,7 @@ def load(self, name: str, index: sympy.Expr):
                 self.disable_vec("not a loop")
                 return var
 
-            if load_dtype not in self.load_supported_dtypes and (
+            if load_dtype not in self.supported_dtypes and (
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp.py:2895]: Also check things in `constant` and `index_expr`?

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: torch/_inductor/codegen/cpp.py (hunk 6 above)
Comment (leslie-fang-intel) [reply torch/_inductor/codegen/cpp.py:2895]: Unify the list in `constant`, `to_dtype`. I didn't see the dtype list in `index_expr`.

Comment (jgong5) [reply torch/_inductor/codegen/cpp.py:2895]: It is a typo.

File: aten/src/ATen/cpu/vec/vec_mask.h (hunk 5 above)
Comment (leslie-fang-intel) [reply aten/src/ATen/cpu/vec/vec_mask.h:137]: There are several places in cpp.py using `store` instead of `store_mask`, such as:

- https://github.com/pytorch/pytorch/blob/ec00daf4f1b09b830a2dfc300a859b5d6f9f39e5/torch/_inductor/codegen/cpp.py#L2449-L2453
- https://github.com/pytorch/pytorch/blob/ec00daf4f1b09b830a2dfc300a859b5d6f9f39e5/torch/_inductor/codegen/cpp.py#L2343
- https://github.com/pytorch/pytorch/blob/ec00daf4f1b09b830a2dfc300a859b5d6f9f39e5/torch/_inductor/codegen/cpp.py#L2815

I afraid we need to check the dtype in all the places and change `store` to `store_mask`?

Comment (jgong5) [reply aten/src/ATen/cpu/vec/vec_mask.h:137]: oh, that's good point. I was just concerned that the name might be confusing with "masked store" semantics. Perhaps, we'd better rename `storeu` with `masked_storeu`..

**Review by jgong5 — APPROVED** [review]

**Comment by leslie-fang-intel** [comment]
Hi @jgong5 @CaoE, the `blendv` for `VecMask` have been added. Please kindly help to take a look again.

**Review by jgong5 — APPROVED** [review]

**Comment by leslie-fang-intel** [comment]
Hi @lezcano, could you help to take a look of this PR?

**Comment by leslie-fang-intel** [comment]
Hi @peterbell10, Could you help to take a look of this PR?

**Review by lezcano — APPROVED** [review]
Sorry, I was on PTO after today. I'm trying to catch-up on my PRs

**Comment by leslie-fang-intel** [comment]
> Sorry, I was on PTO after today. I'm trying to catch-up on my PRs

Thanks for looking into this PR and enjoy the time off.
